src/client/java/net/neoforged/neoforge/client/neo3d/validation/ValidationGpuDevice.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.client.neo3d.validation;
+
+import com.mojang.blaze3d.buffers.GpuBuffer;
+import com.mojang.blaze3d.pipeline.CompiledRenderPipeline;
+import com.mojang.blaze3d.pipeline.RenderPipeline;
+import com.mojang.blaze3d.shaders.ShaderType;
+import com.mojang.blaze3d.systems.CommandEncoder;
+import com.mojang.blaze3d.systems.GpuDevice;
+import com.mojang.blaze3d.textures.GpuTexture;
+import com.mojang.blaze3d.textures.GpuTextureView;
+import com.mojang.blaze3d.textures.TextureFormat;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+import net.minecraft.resources.ResourceLocation;
+import net.neoforged.neoforge.client.neo3d.GpuDeviceFeatures;
+import net.neoforged.neoforge.client.neo3d.GpuDeviceProperties;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+public class ValidationGpuDevice implements GpuDevice {
```

Comment:
This whole thing might be better as patches in the actual GL device, no? It would reduce the size of the PR by a lot as I understand it, by making the changes a lot more targeted.

Reply:
No, it can't be. This is not targeting validating usage for _just_ the `GlDevice`, but _any_ `GpuDevice` implementation. Having it be its own layer also checks against people reaching behind the interface's back to get to the default `GlDevice`. I also expect it to grow past what would be manageable in patches.

Reply:
Agreed this makes sense to have seperate. The PR is larger, but maintanance wise this should be easier to track as new methods added to GpuDevice by mojang should trigger a compiler error.